# <자바스크립트 학습 내용 정리>
- ES6 향상된 객체 문법 
```
var adress = 'seoul';                           /*
var members = {};                                   각각의 키(속성명)은 변수명으로 정의하고 값은 해당 변수에 할당 
var addFamily = function(age, name, role) {
   this.members[role] = {age, name};
};
var getHeadcount = function() {
    return Object.keys(this.members).length;    */
};

var family = {adress, members, addFamily, getHeadcount};  //famlily객체 선언

family.addFamily(30, 'chloe', 'aunt');                    /*
family.addFamily(3, 'lyn', 'niece');                        familiy객체 속성 addFamily함수를 호출하여 members추가 
family.addFamily(10, 'cola', 'cat');                      */

console.log(family.getHeadcount());                       
```
```
var obj = {};                           //객체를 변수에 할당 
for (var i = 0; i < 4; i++) {           /*
    obj['key'+i] = i;                   객체 obj에 속성접근자[]를 활용하여 속성명이 key0,key1,key2,key3이 되도록 정의 
}                                       */
console.log(obj);

var profile = 'chloe:30';               //변수 profile에 문자열'chloe:30' 할당 
var person = {                                      /*
    [profile] : true,                                 profile변수를 person객체의 키로 사용, profile변수를 문자열을 나누어 계산된 값을
    [profile.split(':')[0]]: profile.split(':')[1]      각각 key, value로 사용
};
console.log(person);                                */
```
- 함수 정리하기
```
var a = function(name) {     //함수 정의하면서 변수 a에 할당 
    console.log(name+"!!");
}

function b(name) {
    console.log(name+"!!");  //일방적인 함수 선언 방식 
}
a('hi');                     //'hi!!'출력
b('hi');                     //'hi!!'출력
```
- 예외 처리하기 
자바스크립트는 에러 발생시 그 즉시 중단되기 때문에 이를 대비해서 예외 처리기가 반드시 필요. 예외 처리시 에러 종류는 크게 두가지로 나뉘며, 1)예상치 못한 에러, 2)개발자가 의도한 에러가 있음. 개발자가 의도한 에러는 코드에서 잘못될 가능성을 예상하고 발생시키는 에러를 의미하며 이 경우에는 **throw문**을 사용 throw문은 고의로 에러를 발새시켜 예외 상황을 알리는 역할을 함. throw문이 실행되면 실행되고 있던 블록을 빠져나감. 
```
function checkNumber(val) {
    if (typeof val !== 'number') throw '유효하지 않은 값';
    console.log('숫자형 값');
}

checkNumber(100)            // '숫자형 값'
checkNumber('100')          // '유효하지 않은 값'
```
throw문은 예외 상황을 미리파악하고 에러를 발생시켜 코드가 실행되지 않도록 하며 프로그램 중단은 막을수 없음. 이에 대한 대응책으로 **try-catch-finally문**이 있음.
```
function checkNumber(val) {
    if (typeof val !== 'number') throw '유효하지 않은 값';
    console.log('숫자형 값');
}

try {
    checkNumber(100);                   /*
    checkNumber('100');                    try문 안에서 에러 발생 시 catch 블록으로 전달 
} catch (e) {                               변수 e에 에러 정보가 전달 
    {console.log(`에러발생 >> ${e}`);         콘솔로 에러 정보 출력
} finally {                                코드가 모두 처리되고 난뒤 마지막에 실행되며 에러여부와 상관없이 실행 됨.   
    console.log('완료');                 */
}
```

- arguments 객체 
argument = 전달인자. 매개 변수가 함수 선언 시 작성되는 변수라면, 전달 인자는 함수가 호출 될때 전달되는 값. 
```
function sum() {
    var total = 0;
    for (var i = 0; i < arguments.length; i++) {
        total += arguments[i];
    }
    return total; 
} 

var sumOf1to3 = sum(1,2,3)
console.log(sumOf1to3);
```

-함수 기본 매개변수 처리
```
function drawChart(width=100, height=150) {          //매개변수로 width와 height 선언 ,전달인자가 없을시 기본으로 할당될 값 각 100,150
    console.log(`${width} X ${height} 차트`);
}
drawChart(120);                                     //draChart호출시 인자로 120 전달, width는 120할당, height는 전달인자가 없이 기본값 150 할당
drawChart(200,300);                                 //인자로 200,300 전달 width 200할당, height 300할당 

```
-함수의 나머지 매개변수 
:나머지 매개변수는 매개변수를 정의할 때 정해지지 않은 매개 변수를 정의 할 수 있게 함. 나머지 매개 변수는 배열.
```
function sum(a, b, ...args) {               //sum함수를 a,b매개변수와 나머지 매개변수로 정의
    var total = a+b;
    for (var i=0; i<args.length; i++) {     /* 나머지 매개변수 배열의
        total += args[i];                    값 total 변수에 할당 
    }                                       */
    return total;
}
console.log(sum(1,2,3,4));                  // '10'출력
console.log(sum(1,2,3,4,5,6));              // '21'출력
```